rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RemoteFetchProcessor.java

```diff
@@ -0,0 +1,201 @@
+{
+  private Configuration conf;
+  BookKeeper bookKeeper;
+  private Queue<FetchRequest> processQueue = null;
+  private ConcurrentMap<String, RangeSet<Long>> rangeMap = null;
+  private ConcurrentMap<String, FileMetadataRequest> fetchRequestMap = null;
+  private Runnable runnableStatusUpdateTask = null;
+  private ScheduledExecutorService processService;
+
+  private static DirectBufferPool bufferPool = new DirectBufferPool();
+
+  int diskReadBufferSize;
+  int numRemoteFetchThreads;
+  int remoteFecthThreadInitialDelay;
+  int remoteFetchThreadInterval;
+  int processThreadInitalDelay;
+  int processThreadInterval;
+
+  private static final Log log = LogFactory.getLog(RemoteFetchProcessor.class);
+
+  public RemoteFetchProcessor(BookKeeper bookKeeper, Configuration conf)
+  {
+    this.bookKeeper = bookKeeper;
+    this.conf = conf;
+    this.processQueue = new ConcurrentLinkedQueue<FetchRequest>();
+    this.rangeMap = new ConcurrentHashMap<String, RangeSet<Long>>();
+    this.fetchRequestMap = new ConcurrentHashMap<String, FileMetadataRequest>();
+
+    this.diskReadBufferSize = CacheConfig.getDiskReadBufferSizeDefault(conf);
+    this.numRemoteFetchThreads = CacheConfig.getNumRemoteFetchThreads(conf);
+    this.remoteFecthThreadInitialDelay = CacheConfig.getRemoteFetchThreadInitalDelayInMS(conf);
+    this.remoteFetchThreadInterval = CacheConfig.getRemoteFetchThreadIntervalInMS(conf);
+    this.processThreadInitalDelay = CacheConfig.getProcessThreadInitialDelayInMS(conf);
+    this.processThreadInterval = CacheConfig.getProcessThreadIntervalInMS(conf);
+
+    this.processService = MoreExecutors.getExitingScheduledExecutorService(
+      new ScheduledThreadPoolExecutor(CacheConfig.getNumRemoteFetchThreads(conf)));
+
+    runnableStatusUpdateTask = new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        try {
+          processRemoteFetchRequest();
+        }
+        catch (Exception ex) {
+          log.error(ex);
+        }
+      }
+    };
+    processService.scheduleWithFixedDelay(runnableStatusUpdateTask, remoteFecthThreadInitialDelay,
+        remoteFetchThreadInterval, TimeUnit.MILLISECONDS);
+  }
+
+  public void addToProcessQueue(String remotePath, long offset, int length, long fileSize, long lastModified)
+  {
+    FetchRequest request = new FetchRequest(remotePath, offset, length, fileSize, lastModified);
+    processQueue.add(request);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception
+  {
+    if (!processQueue.isEmpty()) {
+      FetchRequest request = processQueue.remove();
+      log.info("Processing Request : RemotePath - " + request.getRemotePath() + " Offset : " +
+              request.getOffset() + " Length : " + request.getLength());
+
+      if (!rangeMap.containsKey(request.getRemotePath())) {
+        RangeSet<Long> rangeSet = TreeRangeSet.create();
+        rangeMap.putIfAbsent(request.getRemotePath(), rangeSet);
```

Comment:
I think there could be a race between runOneIteration and processRemoteFetchRequest:
- runOneIteration creates empty entry in rangeMap
- Meanwhile processRemoteFetchRequest runs and processes that entry and removes it too
- runOneIteration continues and tries to read the non existent entry

All the more reason to get rid of multiple scheduled services competing for same variables

Reply:
I agree with that. As per @vrajat 's suggestion I am working on fixing this.